src/redis-cli.c

```diff
@@ -418,92 +425,273 @@ static sds cliVersion(void) {
-        tmp.type = CLI_HELP_COMMAND;
-        tmp.org = &commandHelp[i];
-        helpEntries[pos++] = tmp;
-    }
-}
-
-/* cliInitHelp() setups the helpEntries array with the command and group
- * names from the help.h file. However the Redis instance we are connecting
- * to may support more commands, so this function integrates the previous
- * entries with additional entries obtained using the COMMAND command
- * available in recent versions of Redis. */
-static void cliIntegrateHelp(void) {
+    helpEntry *next;
     if (cliConnect(CC_QUIET) == REDIS_ERR) return;
 
-    redisReply *reply = redisCommand(context, "COMMAND");
-    if(reply == NULL || reply->type != REDIS_REPLY_ARRAY) return;
-
-    /* Scan the array reported by COMMAND and fill only the entries that
-     * don't already match what we have. */
-    for (size_t j = 0; j < reply->elements; j++) {
-        redisReply *entry = reply->element[j];
-        if (entry->type != REDIS_REPLY_ARRAY || entry->elements < 4 ||
-            entry->element[0]->type != REDIS_REPLY_STRING ||
-            entry->element[1]->type != REDIS_REPLY_INTEGER ||
-            entry->element[3]->type != REDIS_REPLY_INTEGER) return;
-        char *cmdname = entry->element[0]->str;
-        int i;
-
-        for (i = 0; i < helpEntriesLen; i++) {
-            helpEntry *he = helpEntries+i;
-            if (!strcasecmp(he->argv[0],cmdname))
-                break;
-        }
-        if (i != helpEntriesLen) continue;
-
-        helpEntriesLen++;
-        helpEntries = zrealloc(helpEntries,sizeof(helpEntry)*helpEntriesLen);
-        helpEntry *new = helpEntries+(helpEntriesLen-1);
-
-        new->argc = 1;
-        new->argv = zmalloc(sizeof(sds));
-        new->argv[0] = sdsnew(cmdname);
-        new->full = new->argv[0];
-        new->type = CLI_HELP_COMMAND;
-        sdstoupper(new->argv[0]);
-
-        struct commandHelp *ch = zmalloc(sizeof(*ch));
-        ch->name = new->argv[0];
-        ch->params = sdsempty();
-        int args = llabs(entry->element[1]->integer);
-        args--; /* Remove the command name itself. */
-        if (entry->element[3]->integer == 1) {
-            ch->params = sdscat(ch->params,"key ");
-            args--;
-        }
-        while(args-- > 0) ch->params = sdscat(ch->params,"arg ");
-        if (entry->element[1]->integer < 0)
-            ch->params = sdscat(ch->params,"...options...");
-        ch->summary = "Help not available";
-        ch->group = 0;
-        ch->since = "not known";
-        new->org = ch;
+    redisReply *commandTable = redisCommand(context, "COMMAND DOCS");
+    if (commandTable == NULL || commandTable->type == REDIS_REPLY_ERROR) {
+        /* New COMMAND DOCS not supported - generate help from COMMAND instead. */
+        freeReplyObject(commandTable);
+        cliInitHelpFromCommand();
+        return;
+    };
+    if (commandTable->type != (config.resp3 ? REDIS_REPLY_MAP : REDIS_REPLY_ARRAY)) return;
```

Comment:
ohh, does this mean that the other assertions down the line are not reachable?
still i think we shouldn't look at `config.resp3`, and instead just be able to handle both types silently.

Reply:
Basically, the effect of this `return` is that if the reply from `COMMAND DOCS` is insanely wrong, it will skip initializing the help data and run without help or hints. If it gets past this, then any failed assertion will crash the CLI.

Reply:
Maybe all the asserts in the help initialization should just silently abort loading the help data and return control to the command line?

Reply:
i'm ok with both the asserts and `return`, just don't think we need to use `config.resp3`.